Core/include/Acts/Utilities/GridAccessHelpers.hpp

```diff
@@ -92,62 +93,164 @@ typename grid_type::point_t accessLocal(
   return accessed;
 }
 
-/// Helper method to convert:
-/// 1-D grids:
-/// * [ 0 - n ]
-/// * [ 0 - m ]
-/// 2-D grid:
-/// * [ 0 - n ] x [ 0, m ]
-///
-/// To a local position that can be looked up via the accessLocal(...) method
-///
-/// @param grid is the grid, we need the axes
-/// @param bin0 is the first bin
-/// @param bin1 is the second bin
-/// @param laccess the local accessors for 1D grid case
-///
-/// @note this method is not very fast, it shall not be used for performance
-///      critical code
-///
-/// @return a local position that can be looked up via accessLocal(...)
-template <typename grid_type>
-Vector2 toLocal(const grid_type& grid, std::size_t bin0, std::size_t bin1,
-                std::size_t laccess = 0u) {
-  Vector2 lposition{};
-  // One-dimensional case, needs decision which one is assigned
-  if constexpr (grid_type::DIM == 1u) {
-    if (laccess > 1u) {
+}  // namespace GridAccessHelpers
+
+namespace GridAccess {
+
+/// Interface class for owning delegate
+class IGlobalToGridLocal {
+ public:
+  virtual ~IGlobalToGridLocal() = default;
+};
+
+/// Interface class for owning delegate
+class IBoundToGridLocal {
+ public:
+  virtual ~IBoundToGridLocal() = default;
+};
+
+template <typename global_to_grid_local_t>
+class Affine3Transformed final : public IGlobalToGridLocal {
+ public:
+  using grid_local_t = typename global_to_grid_local_t::grid_local_t;
+
+  /// The global to local transformation
+  global_to_grid_local_t globalToGridLocal;
+
+  /// The transformation matrix
+  Transform3 transform;
+
+  /// Constructor
+  ///
+  /// @param g2gl is the global to grid local transformation
+  /// @param t is the transformation matrix
+  Affine3Transformed(global_to_grid_local_t g2gl, const Transform3& t)
+      : globalToGridLocal(std::move(g2gl)), transform(t) {}
+
+  /// Transform in to the local frame, then the grid local position
+  ///
+  /// @param position is the global position
+  ///
+  /// @return the grid position
+  typename global_to_grid_local_t::grid_local_t toGridLocal(
+      const Vector3& position) const {
+    return globalToGridLocal.toGridLocal(transform * position);
+  }
+};
+
+/// @brief A global (potentially castet) sub space of a global
+/// position
+/// @tparam ...Args
+template <BinningValue... Args>
+class GlobalSubspace final : public IGlobalToGridLocal {
+ public:
+  using grid_local_t = std::array<ActsScalar, sizeof...(Args)>;
+
+  // Constructor with sanity checks
+  GlobalSubspace() {
+    if constexpr (sizeof...(Args) == 0) {
+      throw std::invalid_argument(
+          "GlobalSubspace: cannot have an empty binning value list.");
+    }
+
+    if constexpr (sizeof...(Args) > 3) {
+      throw std::invalid_argument(
+          "GlobalSubspace: cannot have more than 3 binning values.");
+    }
```

Comment:
these should be static asserts outside the constructor which would cleanly separate runtime and compiletime errors

Reply:
The point here is - that it should not be possible to event construct objects of this type ... hence, I would argue the code shouldn't compile ?!

Reply:
Changed.